{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d8d5d705-5137-40d9-b075-db8186a3b1a1",
   "metadata": {},
   "source": [
    "# import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d2440ce3-d7d8-40cf-8f87-68bf81ca9883",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# !pip install pyensembl\n",
    "from pyensembl import EnsemblRelease\n",
    "from scipy.stats import poisson"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0f7e4ba-b786-412e-8125-d053f776a262",
   "metadata": {},
   "source": [
    "# Load Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "843c513f-16bb-4fbd-92d3-5f0c602ac220",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load File\n",
    "DATA_PATH = Path('../../Data')\n",
    "cnv = pd.read_csv(Path(DATA_PATH, \"processed/filtered_cnv_status.tsv\"), sep = '\\t').fillna(\"NaN\")\n",
    "expression = pd.read_csv(Path(DATA_PATH, 'processed/TCGA.HNSC.expression_log_tumor.txt.zip'), sep = '\\t').fillna(\"NaN\")\n",
    "metadata = pd.read_csv(Path(DATA_PATH, 'Original/TCGA.HNSC.metadata.txt'), sep = '\\t').fillna(\"NaN\")\n",
    "metadata_processed = pd.read_csv(Path(DATA_PATH, 'processed/processed_metadata.txt'), sep = '\\t').fillna(\"NaN\")\n",
    "mutations = pd.read_csv(Path(DATA_PATH, 'Original/TCGA.HNSC.mutations.txt'), sep = '\\t').fillna(\"NaN\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0463e6bd-da35-4ee7-8a9b-8493857c9f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OS overall survival 0 still alive, 1 dead\n",
    "# DSS disease specific survival, 346,129\n",
    "# DFI disease free interval, NaN 377, 95, 28\n",
    "# PFI Progression free interval, 306,194"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ddd12f8-b083-4528-b4ff-f0c016db4f24",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Temps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f7e4cf11-52e1-4aa5-8fd5-fc475d27d953",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PFI\n",
       "0.0    306\n",
       "1.0    194\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.PFI.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "26b6eb58-f4fd-40e5-85d7-870685acf4af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['patient_id', 'age', 'OS', 'OS_time', 'DSS', 'DSS_time', 'DFI',\n",
       "       'DFI_time', 'gender', 'tumor_site', 'tumor_class', 'tumor_stage',\n",
       "       'hpv_status'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4407adba-b7e1-4f44-965b-341a558298f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hpv_status\n",
       "NaN         375\n",
       "negative     68\n",
       "positive     44\n",
       "unknown      13\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed.hpv_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9cb4e8aa-2809-4d8f-a2db-cb5e23ceb674",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tumor_site\n",
       "tongue                                                                  149\n",
       "Larynx                                                                  113\n",
       "mouth                                                                    94\n",
       "Other and ill-defined sites in lip, oral cavity and pharynx              67\n",
       "Tonsil                                                                   38\n",
       "Oropharynx                                                               12\n",
       "Gum                                                                      11\n",
       "Hypopharynx                                                               8\n",
       "Palate                                                                    4\n",
       "Lip                                                                       3\n",
       "Bones, joints and articular cartilage of other and unspecified sites      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed.tumor_site.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a01e80c7-d0bd-447e-824a-329602d8c143",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hpv_status\n",
       "NaN         19\n",
       "positive    19\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed[metadata_processed.tumor_site==\"Tonsil\"].hpv_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "37e6c729-5951-4dae-91f7-d98e5199ed82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hpv_status\n",
       "NaN         8\n",
       "negative    3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed[metadata_processed.tumor_site==\"Gum\"].hpv_status.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "87cf2eec-18da-4bcd-b52c-db87c394d8c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tumor_site\n",
       "Tonsil         19\n",
       "tongue         12\n",
       "Oropharynx      4\n",
       "Larynx          4\n",
       "mouth           4\n",
       "Hypopharynx     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed[metadata_processed.hpv_status==\"positive\"].tumor_site.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e0963f07-fbe2-4e3b-b176-29a72c308518",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tumor_site\n",
       "tongue                                                         25\n",
       "Larynx                                                         16\n",
       "mouth                                                          10\n",
       "Other and ill-defined sites in lip, oral cavity and pharynx     8\n",
       "Oropharynx                                                      3\n",
       "Gum                                                             3\n",
       "Hypopharynx                                                     2\n",
       "Palate                                                          1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed[metadata_processed.hpv_status==\"negative\"].tumor_site.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "512bd37e-e063-4308-92ef-e4287a5bba0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tumor_stage\n",
       "NaN          19\n",
       "Stage IVA    13\n",
       "Stage III     5\n",
       "Stage II      5\n",
       "Stage I       2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed[metadata_processed.hpv_status==\"positive\"].tumor_stage.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8d8d52c6-fea7-485f-8f25-6d962219f24e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tumor_stage\n",
       "Stage IVA    26\n",
       "NaN          25\n",
       "Stage III     7\n",
       "Stage II      4\n",
       "Stage I       3\n",
       "Stage IVB     2\n",
       "Stage 0       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_processed[metadata_processed.hpv_status==\"negative\"].tumor_stage.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66c2771b-650b-4696-a249-a5636a608e30",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ceabe038-b818-458c-bcd1-cc3bcd057b78",
   "metadata": {},
   "source": [
    "# Wenshu's Analysis Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f14cd1b8-171c-49a9-831f-afa1e40ff413",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from lifelines import KaplanMeierFitter, CoxPHFitter\n",
    "from lifelines.statistics import multivariate_logrank_test\n",
    "import scipy.stats as stats\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "174ffaed-523d-4ae3-a665-a39fa7c1ff56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>age</th>\n",
       "      <th>OS</th>\n",
       "      <th>OS_time</th>\n",
       "      <th>DSS</th>\n",
       "      <th>DSS_time</th>\n",
       "      <th>DFI</th>\n",
       "      <th>DFI_time</th>\n",
       "      <th>tumor_site</th>\n",
       "      <th>tumor_class</th>\n",
       "      <th>...</th>\n",
       "      <th>ATP2A3|489</th>\n",
       "      <th>SELM|140606</th>\n",
       "      <th>H2AFV|94239</th>\n",
       "      <th>SLC2A3|6515</th>\n",
       "      <th>LMAN1|3998</th>\n",
       "      <th>PPP1R12B|4660</th>\n",
       "      <th>PPP1R14B|26472</th>\n",
       "      <th>ARPC5|10092</th>\n",
       "      <th>COX6A2|1339</th>\n",
       "      <th>gender_MALE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-4P-AA8J</td>\n",
       "      <td>66.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tongue</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>9.727429</td>\n",
       "      <td>11.156279</td>\n",
       "      <td>11.419081</td>\n",
       "      <td>9.497083</td>\n",
       "      <td>10.855484</td>\n",
       "      <td>9.815900</td>\n",
       "      <td>11.754474</td>\n",
       "      <td>12.414751</td>\n",
       "      <td>8.084957</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-BA-4074</td>\n",
       "      <td>69.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>462.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>462.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tongue</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>6.898595</td>\n",
       "      <td>12.697296</td>\n",
       "      <td>13.755680</td>\n",
       "      <td>9.809604</td>\n",
       "      <td>11.520614</td>\n",
       "      <td>6.168125</td>\n",
       "      <td>10.434764</td>\n",
       "      <td>13.869150</td>\n",
       "      <td>3.302144</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-BA-4075</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>283.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>283.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tongue</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>6.906734</td>\n",
       "      <td>9.920316</td>\n",
       "      <td>13.083938</td>\n",
       "      <td>10.478466</td>\n",
       "      <td>10.577429</td>\n",
       "      <td>7.496279</td>\n",
       "      <td>11.715151</td>\n",
       "      <td>13.509293</td>\n",
       "      <td>5.226759</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-BA-4076</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>415.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>415.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Larynx</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>7.755809</td>\n",
       "      <td>10.376223</td>\n",
       "      <td>12.205497</td>\n",
       "      <td>8.382308</td>\n",
       "      <td>11.566972</td>\n",
       "      <td>7.040640</td>\n",
       "      <td>10.446804</td>\n",
       "      <td>12.739734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-BA-4077</td>\n",
       "      <td>45.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1134.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1134.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tongue</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>9.451685</td>\n",
       "      <td>10.593195</td>\n",
       "      <td>11.854339</td>\n",
       "      <td>8.736094</td>\n",
       "      <td>11.699499</td>\n",
       "      <td>8.950538</td>\n",
       "      <td>10.177270</td>\n",
       "      <td>12.495920</td>\n",
       "      <td>3.342469</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>TCGA-UF-A7JT</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>993.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>993.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouth</td>\n",
       "      <td>metastasis</td>\n",
       "      <td>...</td>\n",
       "      <td>9.671928</td>\n",
       "      <td>8.984783</td>\n",
       "      <td>11.415145</td>\n",
       "      <td>9.160953</td>\n",
       "      <td>10.359255</td>\n",
       "      <td>7.433477</td>\n",
       "      <td>11.747996</td>\n",
       "      <td>12.536973</td>\n",
       "      <td>1.846955</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>TCGA-UF-A7JV</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hypopharynx</td>\n",
       "      <td>recurrence</td>\n",
       "      <td>...</td>\n",
       "      <td>8.093967</td>\n",
       "      <td>9.105492</td>\n",
       "      <td>11.019445</td>\n",
       "      <td>11.205708</td>\n",
       "      <td>11.971817</td>\n",
       "      <td>8.135986</td>\n",
       "      <td>11.960310</td>\n",
       "      <td>12.378108</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>TCGA-UP-A6WW</td>\n",
       "      <td>58.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>518.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>518.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tongue</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>10.797110</td>\n",
       "      <td>11.345161</td>\n",
       "      <td>12.408051</td>\n",
       "      <td>8.642936</td>\n",
       "      <td>12.013982</td>\n",
       "      <td>10.655271</td>\n",
       "      <td>10.748252</td>\n",
       "      <td>12.097133</td>\n",
       "      <td>3.644122</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>TCGA-WA-A7GZ</td>\n",
       "      <td>58.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>625.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>625.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>625.0</td>\n",
       "      <td>mouth</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>8.438808</td>\n",
       "      <td>7.957195</td>\n",
       "      <td>11.182196</td>\n",
       "      <td>8.275449</td>\n",
       "      <td>11.835995</td>\n",
       "      <td>9.844485</td>\n",
       "      <td>11.088861</td>\n",
       "      <td>11.589548</td>\n",
       "      <td>5.890332</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>TCGA-WA-A7H4</td>\n",
       "      <td>69.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>443.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>443.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>443.0</td>\n",
       "      <td>tongue</td>\n",
       "      <td>primary</td>\n",
       "      <td>...</td>\n",
       "      <td>10.911751</td>\n",
       "      <td>10.279599</td>\n",
       "      <td>11.542340</td>\n",
       "      <td>8.136730</td>\n",
       "      <td>11.328400</td>\n",
       "      <td>10.781868</td>\n",
       "      <td>11.075513</td>\n",
       "      <td>12.062262</td>\n",
       "      <td>7.856544</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>456 rows × 2013 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       patient_id   age   OS  OS_time  DSS  DSS_time  DFI  DFI_time  \\\n",
       "0    TCGA-4P-AA8J  66.0  0.0    102.0  0.0     102.0  NaN       NaN   \n",
       "1    TCGA-BA-4074  69.0  1.0    462.0  1.0     462.0  NaN       NaN   \n",
       "2    TCGA-BA-4075  49.0  1.0    283.0  1.0     283.0  NaN       NaN   \n",
       "3    TCGA-BA-4076  39.0  1.0    415.0  1.0     415.0  NaN       NaN   \n",
       "4    TCGA-BA-4077  45.0  1.0   1134.0  1.0    1134.0  NaN       NaN   \n",
       "..            ...   ...  ...      ...  ...       ...  ...       ...   \n",
       "451  TCGA-UF-A7JT  72.0  1.0    993.0  1.0     993.0  NaN       NaN   \n",
       "452  TCGA-UF-A7JV  62.0  1.0     90.0  1.0      90.0  NaN       NaN   \n",
       "453  TCGA-UP-A6WW  58.0  0.0    518.0  0.0     518.0  NaN       NaN   \n",
       "454  TCGA-WA-A7GZ  58.0  1.0    625.0  0.0     625.0  0.0     625.0   \n",
       "455  TCGA-WA-A7H4  69.0  0.0    443.0  0.0     443.0  0.0     443.0   \n",
       "\n",
       "      tumor_site tumor_class  ...  ATP2A3|489 SELM|140606  H2AFV|94239  \\\n",
       "0         tongue     primary  ...    9.727429   11.156279    11.419081   \n",
       "1         tongue     primary  ...    6.898595   12.697296    13.755680   \n",
       "2         tongue     primary  ...    6.906734    9.920316    13.083938   \n",
       "3         Larynx     primary  ...    7.755809   10.376223    12.205497   \n",
       "4         tongue     primary  ...    9.451685   10.593195    11.854339   \n",
       "..           ...         ...  ...         ...         ...          ...   \n",
       "451        mouth  metastasis  ...    9.671928    8.984783    11.415145   \n",
       "452  Hypopharynx  recurrence  ...    8.093967    9.105492    11.019445   \n",
       "453       tongue     primary  ...   10.797110   11.345161    12.408051   \n",
       "454        mouth     primary  ...    8.438808    7.957195    11.182196   \n",
       "455       tongue     primary  ...   10.911751   10.279599    11.542340   \n",
       "\n",
       "     SLC2A3|6515  LMAN1|3998  PPP1R12B|4660  PPP1R14B|26472  ARPC5|10092  \\\n",
       "0       9.497083   10.855484       9.815900       11.754474    12.414751   \n",
       "1       9.809604   11.520614       6.168125       10.434764    13.869150   \n",
       "2      10.478466   10.577429       7.496279       11.715151    13.509293   \n",
       "3       8.382308   11.566972       7.040640       10.446804    12.739734   \n",
       "4       8.736094   11.699499       8.950538       10.177270    12.495920   \n",
       "..           ...         ...            ...             ...          ...   \n",
       "451     9.160953   10.359255       7.433477       11.747996    12.536973   \n",
       "452    11.205708   11.971817       8.135986       11.960310    12.378108   \n",
       "453     8.642936   12.013982      10.655271       10.748252    12.097133   \n",
       "454     8.275449   11.835995       9.844485       11.088861    11.589548   \n",
       "455     8.136730   11.328400      10.781868       11.075513    12.062262   \n",
       "\n",
       "     COX6A2|1339  gender_MALE  \n",
       "0       8.084957         True  \n",
       "1       3.302144         True  \n",
       "2       5.226759         True  \n",
       "3       0.000000         True  \n",
       "4       3.342469        False  \n",
       "..           ...          ...  \n",
       "451     1.846955        False  \n",
       "452     0.000000        False  \n",
       "453     3.644122         True  \n",
       "454     5.890332         True  \n",
       "455     7.856544         True  \n",
       "\n",
       "[456 rows x 2013 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in metadata and feature matrix\n",
    "# Change to your own paths; feature_mat should have patient_id as rows and features as columns\n",
    "metadata = pd.read_csv(\"../../Data/processed/processed_metadata.txt\", sep=\"\\t\", header=0)\n",
    "metadata = metadata[metadata[\"hpv_status\"] != \"positive\"]\n",
    "metadata = metadata[metadata[\"tumor_site\"] != \"tonsil\"]\n",
    "res_dir = \"Results/\"\n",
    "feature_mat = pd.read_csv(\"../Wenshu/test/sample_data/TCGA.HNSC.expression_feature_matrix.txt\",\n",
    "                          sep=\"\\t\", header=0, index_col=0)\n",
    "\n",
    "features = feature_mat.columns.tolist()\n",
    "\n",
    "df = pd.merge(metadata, feature_mat, on=\"patient_id\")\n",
    "df = pd.get_dummies(df, columns=['gender'], drop_first=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1766a391-ad46-436b-b026-605c3f24c6b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Univariate Cox Proportional Hazards Models\n",
    "def univariate_cox(df, data_type, q_thresh = 0.05):\n",
    "    results = []\n",
    "    for feature in features:\n",
    "        if(data_type in (\"cna\", \"mutation\")):\n",
    "            feat = df[feature].dropna().astype(str)\n",
    "            # one-hot encode this feature only\n",
    "            dummies = pd.get_dummies(feat, prefix=feature, drop_first=True)\n",
    "            cols = ['OS_time','OS'] + list(dummies.columns)\n",
    "\n",
    "            data = pd.concat([df[['OS_time','OS']], dummies], axis=1).loc[:, cols].dropna()\n",
    "            \n",
    "            cph = CoxPHFitter()\n",
    "            cph.fit(data,\n",
    "                    duration_col='OS_time',\n",
    "                    event_col='OS',\n",
    "                    show_progress=False)\n",
    "            \n",
    "            for cov in dummies.columns:\n",
    "                summ = cph.summary.loc[cov]\n",
    "                results.append({\n",
    "                    'feature':   feature,\n",
    "                    'covariate': cov,\n",
    "                    'coef':      summ['coef'],\n",
    "                    'exp(coef)': summ['exp(coef)'],\n",
    "                    'zscore':    summ['z'],\n",
    "                    'p':     summ['p'],\n",
    "                    'lower95%':  summ['exp(coef) lower 95%'],\n",
    "                    'upper95%':  summ['exp(coef) upper 95%']\n",
    "                })\n",
    "        else:\n",
    "            data = df[[feature, 'OS_time', 'OS']].dropna()\n",
    "\n",
    "            cph = CoxPHFitter()\n",
    "            cph.fit(data, duration_col='OS_time', event_col='OS', show_progress=False)\n",
    "            summ = cph.summary.loc[feature]\n",
    "\n",
    "            results.append({\n",
    "                'feature':     feature,\n",
    "                'coef':        summ['coef'],\n",
    "                'exp(coef)':   summ['exp(coef)'],\n",
    "                'zscore':      summ['z'],\n",
    "                'p':           summ['p'],\n",
    "                'lower 95%':   summ['exp(coef) lower 95%'],\n",
    "                'upper 95%':   summ['exp(coef) upper 95%']\n",
    "            })\n",
    "\n",
    "    res_df = pd.DataFrame(results).set_index('feature')\n",
    "    res_df[\"q\"] = multipletests(res_df['p'], method='fdr_bh')[1]\n",
    "    res_df = res_df.sort_values('q', ascending=True)\n",
    "    res_df.to_csv(f'{res_dir}{data_type}_univariate_cox_results.tsv', sep=\"\\t\", index=True, header=True)\n",
    "    sig_features = res_df[res_df['q'] < q_thresh]\n",
    "    sig_features.to_csv(f'{res_dir}{data_type}_univariate_cox_significant_features.tsv', sep=\"\\t\", index=True, header=True)\n",
    "    return sig_features\n",
    "\n",
    "# Multivariate Cox Proportional Hazards Models\n",
    "def multivariate_cox(df, data_type, q_thresh = 0.05):\n",
    "    results = []\n",
    "    for feature in features:\n",
    "        if(data_type in (\"cna\", \"mutation\")):\n",
    "            feat = df[feature].dropna().astype(str)\n",
    "            # one-hot encode this feature only\n",
    "            dummies = pd.get_dummies(feat, prefix=feature, drop_first=True)\n",
    "            cols = ['OS_time','OS','age', 'tumor_stage'] \\\n",
    "                + [c for c in df if c.startswith('gender_')] \\\n",
    "                + list(dummies.columns)\n",
    "\n",
    "            data = pd.concat([df[['OS_time','OS','age', 'tumor_stage'] + \n",
    "                             [c for c in df if c.startswith('gender_')]],\n",
    "                            dummies], axis=1).loc[:, cols].dropna()\n",
    "            \n",
    "            cph = CoxPHFitter()\n",
    "            cph.fit(data, duration_col='OS_time', event_col='OS', show_progress=False)\n",
    "            \n",
    "            for cov in dummies.columns:\n",
    "                summ = cph.summary.loc[cov]\n",
    "                results.append({\n",
    "                    'feature':   feature,\n",
    "                    'covariate': cov,\n",
    "                    'coef':      summ['coef'],\n",
    "                    'exp(coef)': summ['exp(coef)'],\n",
    "                    'zscore':    summ['z'],\n",
    "                    'p':     summ['p'],\n",
    "                    'lower95%':  summ['exp(coef) lower 95%'],\n",
    "                    'upper95%':  summ['exp(coef) upper 95%']\n",
    "                })\n",
    "        else:\n",
    "            cols = ['OS_time','OS','age', 'tumor_stage'] \\\n",
    "                + [c for c in df.columns if c.startswith('gender_')] \\\n",
    "                + [feature]\n",
    "            data = df[cols].dropna()\n",
    "\n",
    "            cph = CoxPHFitter()\n",
    "            cph.fit(data, duration_col='OS_time', event_col='OS', show_progress=False)\n",
    "            summ = cph.summary.loc[feature]\n",
    "\n",
    "            results.append({\n",
    "                'feature':     feature,\n",
    "                'coef':        summ['coef'],\n",
    "                'exp(coef)':   summ['exp(coef)'],\n",
    "                'zscore':      summ['z'],\n",
    "                'p':           summ['p'],\n",
    "                'lower 95%':   summ['exp(coef) lower 95%'],\n",
    "                'upper 95%':   summ['exp(coef) upper 95%']\n",
    "            })\n",
    "\n",
    "    res_df = pd.DataFrame(results).set_index('feature')\n",
    "    res_df[\"q\"] = multipletests(res_df['p'], method='fdr_bh')[1]\n",
    "    res_df = res_df.sort_values('q', ascending=True)\n",
    "    res_df.to_csv(f'{res_dir}{data_type}_multivariate_cox_results.tsv', sep=\"\\t\", index=True, header=True)\n",
    "    sig_features = res_df[res_df['q'] < q_thresh]\n",
    "    sig_features.to_csv(f'{res_dir}{data_type}_multivariate_cox_significant_features.tsv', sep=\"\\t\", index=True, header=True)\n",
    "    return sig_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d598ce8d-6cfa-4fb0-8f78-e6e2604204ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# univariate_results = univariate_cox(df, \"expression\", q_thresh = 0.05)\n",
    "# multivariate_results = multivariate_cox(df, \"expression\", q_thresh = 0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1ae5830-4076-4476-80b1-d9b49eb06a84",
   "metadata": {},
   "source": [
    "# Functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7078f110-44ae-440a-8d85-bead83607d33",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_to_aa = {\n",
    "    'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "    'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M',\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "    'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "    'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4a77c27d-2faf-4d2b-93f2-467428225f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_dn_ds(vclass):\n",
    "    if vclass == 'Silent':\n",
    "        return 'synonymous'\n",
    "    else:\n",
    "        return 'nonsynonymous'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dba26059-544d-4f7f-ba4e-ee2bdebdbea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_substitution(ref, alt):\n",
    "    valid_bases = {'A', 'C', 'G', 'T'}\n",
    "    if ref not in valid_bases or alt not in valid_bases:\n",
    "        return None\n",
    "\n",
    "    if ref in ['C', 'T']:\n",
    "        return ref + \"->\" + alt\n",
    "    else:\n",
    "        # Reverse complement both ref and alt\n",
    "        complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "        norm_ref = complement[ref]\n",
    "        norm_alt = complement[alt]\n",
    "        return norm_ref + \"->\" + norm_alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "68faccc7-0bd0-429e-b848-0b519671452d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dn_ds_ratio(subset):\n",
    "    n_dN = (subset['_dn_ds_mutation_type'] == 'nonsynonymous').sum()\n",
    "    n_dS = (subset['_dn_ds_mutation_type'] == 'synonymous').sum()\n",
    "    if n_dS == 0:\n",
    "        ratio = (n_dN + 1) / (n_dS + 1)  # add pseudocount\n",
    "    else:\n",
    "        ratio = n_dN/n_dS\n",
    "    return ratio "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "21ffa73b-36e5-4f87-92e1-f245423d3137",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_gene_mutations(cds_seq, mutation_rate_dict, codon_to_aa):\n",
    "    results = []\n",
    "    codons = [cds_seq[i:i+3] for i in range(0, len(cds_seq), 3)]\n",
    "\n",
    "    for codon_idx, codon in enumerate(codons):\n",
    "        if len(codon) < 3 or codon not in codon_to_aa:\n",
    "            continue  # skip incomplete or invalid codons\n",
    "\n",
    "        for pos in range(3):  # position within codon\n",
    "            ref_base = codon[pos]\n",
    "\n",
    "            for alt_base in ['A', 'C', 'G', 'T']:\n",
    "                if alt_base == ref_base:\n",
    "                    continue\n",
    "\n",
    "                # Mutate codon\n",
    "                mutated = list(codon)\n",
    "                mutated[pos] = alt_base\n",
    "                mutated_codon = ''.join(mutated)\n",
    "\n",
    "                if mutated_codon not in codon_to_aa:\n",
    "                    continue  # skip invalid codons\n",
    "\n",
    "                # Determine mutation type\n",
    "                original_aa = codon_to_aa[codon]\n",
    "                mutated_aa = codon_to_aa[mutated_codon]\n",
    "                mutation_type = 'synonymous' if original_aa == mutated_aa else 'nonsynonymous'\n",
    "\n",
    "                # Get center position in CDS\n",
    "                genome_pos = codon_idx * 3 + pos\n",
    "                if genome_pos < 1 or genome_pos >= len(cds_seq) - 1:\n",
    "                    continue  # skip edges\n",
    "\n",
    "                # Get trinucleotide context (5' and 3' bases)\n",
    "                context = cds_seq[genome_pos - 1] + \"_\" + cds_seq[genome_pos + 1]\n",
    "                substitution = get_substitution(ref_base, alt_base)\n",
    "\n",
    "                # Get mutation rate\n",
    "                mutation_prob = mutation_rate_dict.get((context, substitution), 0.0)\n",
    "\n",
    "                results.append({\n",
    "                    'codon_index': codon_idx,\n",
    "                    'codon': codon,\n",
    "                    'mutated_codon': mutated_codon,\n",
    "                    'ref_base': ref_base,\n",
    "                    'alt_base': alt_base,\n",
    "                    'position_in_codon': pos,\n",
    "                    'context': context,\n",
    "                    'substitution': substitution,\n",
    "                    'mutation_type': mutation_type,\n",
    "                    'mutation_probability': mutation_prob\n",
    "                })\n",
    "\n",
    "    return results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e28c5ba-3092-4375-9849-444759e77eba",
   "metadata": {},
   "source": [
    "# Preprocessing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "209c9b96-8edd-4248-b13c-3b74e3e662c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. filter to pass: kept high-quality mutations\n",
    "mutations = mutations[mutations['FILTER'] == 'PASS']\n",
    "\n",
    "# 2. remove deprecated vairant callers and missing callers\n",
    "# from mutations['CALLERS'], remove str matching 'RADIA' and 'SOMATICSNIPER'\n",
    "mutations['CALLERS'] = mutations['CALLERS'].replace(to_replace='*', value='')\n",
    "mutations['CALLERS'] = mutations['CALLERS'].replace(to_replace='RADIA', value='')\n",
    "mutations['CALLERS'] = mutations['CALLERS'].replace(to_replace='SOMATICSNIPER', value='')\n",
    "# in mutations['CALLERS'], if 2 '|' are next to each other, replace with ''\n",
    "mutations['CALLERS'] = mutations['CALLERS'].replace(to_replace='\\\\|\\\\|', value='', regex=True)\n",
    "# count the number of '|' in mutations['CALLERS'] and store in a new column 'CALLERS_COUNT_ADJUSTED'\n",
    "mutations['CALLERS_COUNT_ADJUSTED'] = mutations['CALLERS'].str.count('\\\\|')\n",
    "\n",
    "# 3. Filter to consensus mutations\n",
    "mutations = mutations[mutations['CALLERS_COUNT_ADJUSTED'] > 1]\n",
    "\n",
    "# 4. Filter by n_depth >=10\n",
    "mutations = mutations[mutations['n_depth'] >= 10] \n",
    "\n",
    "# 5. Filter by hypermutators > 1000 mutations per patient\n",
    "# Identify patients with >1000 mutations\n",
    "hypermutators = mutations['patient_id'].value_counts()\n",
    "hypermutators = hypermutators[hypermutators > 1000].index\n",
    "# Remove those patients\n",
    "mutations = mutations[~mutations['patient_id'].isin(hypermutators)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "40feb9c4-16dd-4cdb-b461-693526a21d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !!!!! additional step !!!!!\n",
    "# 6. Only SNP mutations should be analyzed\n",
    "mutations = mutations[mutations['Variant_Type'] == 'SNP']\n",
    "\n",
    "# 7. additional column on mutation type and trinucleotide contextx\n",
    "# dn/ds mutation\n",
    "mutations.loc[:, '_dn_ds_mutation_type'] = mutations['Variant_Classification'].apply(classify_dn_ds)\n",
    "# trinucleotide\n",
    "mutations.loc[:, '_trinucleotide'] = mutations['CONTEXT'].str.slice(4, 7)\n",
    "# 16 type context\n",
    "mutations['_16_type_context'] = mutations['CONTEXT'].str[4] + '_' + mutations['CONTEXT'].str[6]\n",
    "# 6 substitution \n",
    "# mutations['_substitution'] = mutations.apply(\n",
    "#     lambda row: get_substitution(row['Reference_Allele'], row['Tumor_Seq_Allele2']), axis=1\n",
    "# )\n",
    "mutations['_substitution'] = mutations['Reference_Allele'] + \"->\" + mutations['Tumor_Seq_Allele2']\n",
    "mutations['_96_class'] = mutations['_16_type_context'] + '__' + mutations['_substitution']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d6222da-971e-44c0-b9de-4eac9330ab2c",
   "metadata": {},
   "source": [
    "# Select Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1109fa66-4cdf-408e-93b8-77f9f06c2643",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR! Session/line number was not unique in database. History logging moved to new session 42\n"
     ]
    }
   ],
   "source": [
    "mutations = mutations[mutations._dn_ds_mutation_type == \"nonsynonymous\"]\n",
    "mutations = mutations[mutations.Variant_Classification != \"Intron\"]\n",
    "mutations = mutations[mutations.Variant_Classification != \"3'UTR\"] # untranslated\n",
    "mutations = mutations[mutations.Variant_Classification != \"5'UTR\"]\n",
    "mutations = mutations[mutations.Variant_Classification != \"RNA\"] # RNA region\n",
    "mutations = mutations[mutations.Variant_Classification != \"3'Flank\"] # flanking region\n",
    "mutations = mutations[mutations.Variant_Classification != \"5'Flank\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3b641f01-3f9b-42a2-bac2-a5302b6617e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 2: Create binary mutation matrix (patients as rows, genes as columns)\n",
    "binary_matrix = mutations.groupby(['patient_id', 'Hugo_Symbol']) \\\n",
    "                         .size().unstack(fill_value=0)\n",
    "\n",
    "# Step 3: Convert counts to 1 (presence) or 0 (absence)\n",
    "binary_matrix = (binary_matrix > 0).astype(int)\n",
    "\n",
    "binary_matrix = binary_matrix.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64dc172e-824e-4e1d-9637-0704307afd3e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9c67dec1-da04-477b-a4e2-7b92b4d4baed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you've accidentally included gene names as index and want patient_id as index:\n",
    "binary_matrix = binary_matrix.set_index('patient_id')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ec10c5da-6b47-4cfe-9104-d2ffb4473785",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AADACL2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-4P-AA8J</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BA-4074</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BA-4075</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BA-4076</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BA-4077</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-UF-A7JT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-UF-A7JV</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-UP-A6WW</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-WA-A7GZ</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-WA-A7H4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>494 rows × 14295 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Hugo_Symbol   A1BG  A1CF  A2M  A2ML1  A4GALT  A4GNT  AAAS  AACS  AADAC  \\\n",
       "patient_id                                                               \n",
       "TCGA-4P-AA8J     0     0    0      0       0      0     0     0      0   \n",
       "TCGA-BA-4074     0     0    0      0       0      0     0     0      0   \n",
       "TCGA-BA-4075     0     0    0      0       0      0     0     0      0   \n",
       "TCGA-BA-4076     0     0    0      0       0      0     0     0      0   \n",
       "TCGA-BA-4077     0     0    0      0       0      0     0     0      0   \n",
       "...            ...   ...  ...    ...     ...    ...   ...   ...    ...   \n",
       "TCGA-UF-A7JT     0     0    0      0       0      0     0     0      0   \n",
       "TCGA-UF-A7JV     0     0    0      0       0      0     0     0      0   \n",
       "TCGA-UP-A6WW     0     0    0      0       0      0     0     0      0   \n",
       "TCGA-WA-A7GZ     0     0    0      0       0      0     0     0      0   \n",
       "TCGA-WA-A7H4     0     0    0      0       0      0     0     0      0   \n",
       "\n",
       "Hugo_Symbol   AADACL2  ...  ZW10  ZWILCH  ZWINT  ZXDA  ZXDB  ZXDC  ZYG11B  \\\n",
       "patient_id             ...                                                  \n",
       "TCGA-4P-AA8J        0  ...     0       0      0     0     0     0       0   \n",
       "TCGA-BA-4074        0  ...     0       0      0     0     0     0       0   \n",
       "TCGA-BA-4075        0  ...     0       0      0     0     0     0       0   \n",
       "TCGA-BA-4076        0  ...     0       0      0     0     0     0       0   \n",
       "TCGA-BA-4077        0  ...     0       0      0     0     0     0       0   \n",
       "...               ...  ...   ...     ...    ...   ...   ...   ...     ...   \n",
       "TCGA-UF-A7JT        0  ...     0       0      0     0     0     0       0   \n",
       "TCGA-UF-A7JV        0  ...     0       0      0     0     0     0       0   \n",
       "TCGA-UP-A6WW        0  ...     0       0      0     0     0     0       0   \n",
       "TCGA-WA-A7GZ        0  ...     0       1      0     0     0     0       0   \n",
       "TCGA-WA-A7H4        0  ...     0       0      0     0     0     0       0   \n",
       "\n",
       "Hugo_Symbol   ZYX  ZZEF1  ZZZ3  \n",
       "patient_id                      \n",
       "TCGA-4P-AA8J    0      0     0  \n",
       "TCGA-BA-4074    0      0     0  \n",
       "TCGA-BA-4075    0      0     0  \n",
       "TCGA-BA-4076    0      0     0  \n",
       "TCGA-BA-4077    0      1     0  \n",
       "...           ...    ...   ...  \n",
       "TCGA-UF-A7JT    0      0     0  \n",
       "TCGA-UF-A7JV    0      0     0  \n",
       "TCGA-UP-A6WW    0      0     0  \n",
       "TCGA-WA-A7GZ    0      0     0  \n",
       "TCGA-WA-A7H4    0      0     0  \n",
       "\n",
       "[494 rows x 14295 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binary_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "25ce88be-343e-4a46-abda-f1838f1aa404",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = binary_matrix.columns.tolist()\n",
    "\n",
    "# Count mutated patients per gene\n",
    "mutated_counts = binary_matrix[features].sum()\n",
    "\n",
    "# Keep genes mutated in at least 5 patients\n",
    "filtered_features = mutated_counts[mutated_counts >= 20].index.tolist()\n",
    "\n",
    "binary_matrix = binary_matrix[filtered_features].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b63d506-9f05-43be-99a4-13bd82d22c10",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "24878747-1c19-47c1-a1c6-57ebe57ea8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge metadata with mutation features\n",
    "df = pd.merge(metadata, binary_matrix, on=\"patient_id\")\n",
    "\n",
    "# One-hot encode gender\n",
    "df = pd.get_dummies(df, columns=['gender'], drop_first=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96c7df87-813d-4570-8d07-b7b09fb3d49b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e537674-213c-4f08-a925-573ba182209a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f214b61c-2c02-433e-bff5-553fc6b37981",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "6cf0667d-0164-4bed-90c6-5148e5e2f014",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/utils/__init__.py:1120: ConvergenceWarning: Column A1BG_1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['OS'].astype(bool)\n",
      ">>> print(df.loc[events, 'A1BG_1'].var())\n",
      ">>> print(df.loc[~events, 'A1BG_1'].var())\n",
      "\n",
      "A very low variance means that the column A1BG_1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1607: ConvergenceWarning: Newton-Raphson convergence completed successfully but norm(delta) is still high, 0.105. This may imply non-unique solutions to the maximum likelihood. Perhaps there is collinearity or complete separation in the dataset?\n",
      "\n",
      "  warnings.warn(\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1679: RuntimeWarning: overflow encountered in exp\n",
      "  scores = weights * exp(dot(X, beta))\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1731: RuntimeWarning: divide by zero encountered in divide\n",
      "  denom = 1.0 / np.array([risk_phi])\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1733: RuntimeWarning: invalid value encountered in multiply\n",
      "  a1 = risk_phi_x_x * denom\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1735: RuntimeWarning: invalid value encountered in multiply\n",
      "  summand = numer * denom[:, None]\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1727: RuntimeWarning: divide by zero encountered in divide\n",
      "  denom = 1.0 / (risk_phi - increasing_proportion * tie_phi)\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1729: RuntimeWarning: invalid value encountered in multiply\n",
      "  a1 = einsum(\"ab,i->ab\", risk_phi_x_x, denom) - einsum(\"ab,i->ab\", tie_phi_x_x, increasing_proportion * denom)\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1727: RuntimeWarning: invalid value encountered in multiply\n",
      "  denom = 1.0 / (risk_phi - increasing_proportion * tie_phi)\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1727: RuntimeWarning: invalid value encountered in subtract\n",
      "  denom = 1.0 / (risk_phi - increasing_proportion * tie_phi)\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1728: RuntimeWarning: invalid value encountered in multiply\n",
      "  numer = risk_phi_x - multiply.outer(increasing_proportion, tie_phi_x)\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1728: RuntimeWarning: invalid value encountered in subtract\n",
      "  numer = risk_phi_x - multiply.outer(increasing_proportion, tie_phi_x)\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1740: RuntimeWarning: divide by zero encountered in log\n",
      "  log_lik = log_lik + dot(x_death_sum, beta) + weighted_average * log(denom).sum()\n",
      "/Users/louxuwen/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/utils/__init__.py:1120: ConvergenceWarning: Column AAMP_1 have very low variance when conditioned on death event present or not. This may harm convergence. This could be a form of 'complete separation'. For example, try the following code:\n",
      "\n",
      ">>> events = df['OS'].astype(bool)\n",
      ">>> print(df.loc[events, 'AAMP_1'].var())\n",
      ">>> print(df.loc[~events, 'AAMP_1'].var())\n",
      "\n",
      "A very low variance means that the column AAMP_1 completely determines whether a subject dies or not. See https://stats.stackexchange.com/questions/11109/how-to-deal-with-perfect-separation-in-logistic-regression.\n",
      "\n",
      "  warnings.warn(dedent(warning_text), ConvergenceWarning)\n"
     ]
    },
    {
     "ename": "ConvergenceError",
     "evalue": "Convergence halted due to matrix inversion problems. Suspicion is high collinearity. Please see the following tips in the lifelines documentation: https://lifelines.readthedocs.io/en/latest/Examples.html#problems-with-convergence-in-the-cox-proportional-hazard-modelMatrix is singular.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLinAlgError\u001b[0m                               Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1530\u001b[0m, in \u001b[0;36mSemiParametricPHFitter._newton_raphson_for_efron_model\u001b[0;34m(self, X, T, E, weights, entries, initial_point, show_progress, step_size, precision, r_precision, max_steps)\u001b[0m\n\u001b[1;32m   1529\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1530\u001b[0m     inv_h_dot_g_T \u001b[38;5;241m=\u001b[39m \u001b[43mspsolve\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[43mh\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43massume_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpos\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_finite\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m   1531\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mValueError\u001b[39;00m, LinAlgError) \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/scipy/linalg/_basic.py:337\u001b[0m, in \u001b[0;36msolve\u001b[0;34m(a, b, lower, overwrite_a, overwrite_b, check_finite, assume_a, transposed)\u001b[0m\n\u001b[1;32m    334\u001b[0m lu, x, info \u001b[38;5;241m=\u001b[39m posv(a1, b1, lower\u001b[38;5;241m=\u001b[39mlower,\n\u001b[1;32m    335\u001b[0m                    overwrite_a\u001b[38;5;241m=\u001b[39moverwrite_a,\n\u001b[1;32m    336\u001b[0m                    overwrite_b\u001b[38;5;241m=\u001b[39moverwrite_b)\n\u001b[0;32m--> 337\u001b[0m \u001b[43m_solve_check\u001b[49m\u001b[43m(\u001b[49m\u001b[43mn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minfo\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    338\u001b[0m rcond, info \u001b[38;5;241m=\u001b[39m pocon(lu, anorm)\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/scipy/linalg/_basic.py:42\u001b[0m, in \u001b[0;36m_solve_check\u001b[0;34m(n, info, lamch, rcond)\u001b[0m\n\u001b[1;32m     41\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;241m0\u001b[39m \u001b[38;5;241m<\u001b[39m info:\n\u001b[0;32m---> 42\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m LinAlgError(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMatrix is singular.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     44\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m lamch \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[0;31mLinAlgError\u001b[0m: Matrix is singular.",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConvergenceError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[93], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m sig_uni \u001b[38;5;241m=\u001b[39m \u001b[43munivariate_cox\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmutation\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mq_thresh\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.05\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m sig_multi \u001b[38;5;241m=\u001b[39m multivariate_cox(df, data_type\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmutation\u001b[39m\u001b[38;5;124m\"\u001b[39m, q_thresh\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.05\u001b[39m)\n",
      "Cell \u001b[0;32mIn[45], line 14\u001b[0m, in \u001b[0;36munivariate_cox\u001b[0;34m(df, data_type, q_thresh)\u001b[0m\n\u001b[1;32m     11\u001b[0m data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([df[[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mOS_time\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mOS\u001b[39m\u001b[38;5;124m'\u001b[39m]], dummies], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\u001b[38;5;241m.\u001b[39mloc[:, cols]\u001b[38;5;241m.\u001b[39mdropna()\n\u001b[1;32m     13\u001b[0m cph \u001b[38;5;241m=\u001b[39m CoxPHFitter()\n\u001b[0;32m---> 14\u001b[0m \u001b[43mcph\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[43m        \u001b[49m\u001b[43mduration_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mOS_time\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[43m        \u001b[49m\u001b[43mevent_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mOS\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     17\u001b[0m \u001b[43m        \u001b[49m\u001b[43mshow_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m cov \u001b[38;5;129;01min\u001b[39;00m dummies\u001b[38;5;241m.\u001b[39mcolumns:\n\u001b[1;32m     20\u001b[0m     summ \u001b[38;5;241m=\u001b[39m cph\u001b[38;5;241m.\u001b[39msummary\u001b[38;5;241m.\u001b[39mloc[cov]\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/utils/__init__.py:56\u001b[0m, in \u001b[0;36mCensoringType.right_censoring.<locals>.f\u001b[0;34m(model, *args, **kwargs)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(function)\n\u001b[1;32m     54\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mf\u001b[39m(model, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mset_censoring_type(model, \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mRIGHT)\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:290\u001b[0m, in \u001b[0;36mCoxPHFitter.fit\u001b[0;34m(self, df, duration_col, event_col, show_progress, initial_point, strata, weights_col, cluster_col, robust, batch_mode, timeline, formula, entry_col, fit_options)\u001b[0m\n\u001b[1;32m    184\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    185\u001b[0m \u001b[38;5;124;03mFit the Cox proportional hazard model to a right-censored dataset. Alias of `fit_right_censoring`.\u001b[39;00m\n\u001b[1;32m    186\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    287\u001b[0m \n\u001b[1;32m    288\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    289\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata \u001b[38;5;241m=\u001b[39m utils\u001b[38;5;241m.\u001b[39m_to_list_or_singleton(utils\u001b[38;5;241m.\u001b[39mcoalesce(strata, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata))\n\u001b[0;32m--> 290\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_model \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    291\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    292\u001b[0m \u001b[43m    \u001b[49m\u001b[43mduration_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    293\u001b[0m \u001b[43m    \u001b[49m\u001b[43mevent_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mevent_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    294\u001b[0m \u001b[43m    \u001b[49m\u001b[43mshow_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mshow_progress\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    295\u001b[0m \u001b[43m    \u001b[49m\u001b[43minitial_point\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_point\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    296\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstrata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstrata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    297\u001b[0m \u001b[43m    \u001b[49m\u001b[43mweights_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mweights_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    298\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcluster_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcluster_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    299\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrobust\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrobust\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    300\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbatch_mode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbatch_mode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    301\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeline\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    302\u001b[0m \u001b[43m    \u001b[49m\u001b[43mformula\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mformula\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    303\u001b[0m \u001b[43m    \u001b[49m\u001b[43mentry_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mentry_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    304\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfit_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfit_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    305\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    306\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:610\u001b[0m, in \u001b[0;36mCoxPHFitter._fit_model\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    608\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_fit_model\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    609\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbaseline_estimation_method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbreslow\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m--> 610\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_model_breslow\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    611\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbaseline_estimation_method \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mspline\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    612\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fit_model_spline(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:623\u001b[0m, in \u001b[0;36mCoxPHFitter._fit_model_breslow\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    619\u001b[0m model \u001b[38;5;241m=\u001b[39m SemiParametricPHFitter(\n\u001b[1;32m    620\u001b[0m     penalizer\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpenalizer, l1_ratio\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39ml1_ratio, strata\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstrata, alpha\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39malpha, label\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_label\n\u001b[1;32m    621\u001b[0m )\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m utils\u001b[38;5;241m.\u001b[39mCensoringType\u001b[38;5;241m.\u001b[39mis_right_censoring(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 623\u001b[0m     \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    624\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m model\n\u001b[1;32m    625\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/utils/__init__.py:56\u001b[0m, in \u001b[0;36mCensoringType.right_censoring.<locals>.f\u001b[0;34m(model, *args, **kwargs)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(function)\n\u001b[1;32m     54\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mf\u001b[39m(model, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mset_censoring_type(model, \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mRIGHT)\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1252\u001b[0m, in \u001b[0;36mSemiParametricPHFitter.fit\u001b[0;34m(self, df, duration_col, event_col, show_progress, initial_point, strata, weights_col, cluster_col, robust, batch_mode, timeline, formula, entry_col, fit_options)\u001b[0m\n\u001b[1;32m   1247\u001b[0m \u001b[38;5;66;03m# this is surprisingly faster to do...\u001b[39;00m\n\u001b[1;32m   1248\u001b[0m X_norm \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[1;32m   1249\u001b[0m     utils\u001b[38;5;241m.\u001b[39mnormalize(X\u001b[38;5;241m.\u001b[39mvalues, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_norm_mean\u001b[38;5;241m.\u001b[39mvalues, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_norm_std\u001b[38;5;241m.\u001b[39mvalues), index\u001b[38;5;241m=\u001b[39mX\u001b[38;5;241m.\u001b[39mindex, columns\u001b[38;5;241m=\u001b[39mX\u001b[38;5;241m.\u001b[39mcolumns\n\u001b[1;32m   1250\u001b[0m )\n\u001b[0;32m-> 1252\u001b[0m params_, ll_, variance_matrix_, baseline_hazard_, baseline_cumulative_hazard_, model \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fit_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1253\u001b[0m \u001b[43m    \u001b[49m\u001b[43mX_norm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1254\u001b[0m \u001b[43m    \u001b[49m\u001b[43mT\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1255\u001b[0m \u001b[43m    \u001b[49m\u001b[43mE\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1256\u001b[0m \u001b[43m    \u001b[49m\u001b[43mweights\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mweights\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1257\u001b[0m \u001b[43m    \u001b[49m\u001b[43mentries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mentries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1258\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfit_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mutils\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcoalesce\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfit_options\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mdict\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1259\u001b[0m \u001b[43m    \u001b[49m\u001b[43minitial_point\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_point\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1260\u001b[0m \u001b[43m    \u001b[49m\u001b[43mshow_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mshow_progress\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1261\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1263\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlog_likelihood_ \u001b[38;5;241m=\u001b[39m ll_\n\u001b[1;32m   1264\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel \u001b[38;5;241m=\u001b[39m model\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1380\u001b[0m, in \u001b[0;36mSemiParametricPHFitter._fit_model\u001b[0;34m(self, X, T, E, weights, entries, fit_options, initial_point, show_progress)\u001b[0m\n\u001b[1;32m   1369\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_fit_model\u001b[39m(\n\u001b[1;32m   1370\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1371\u001b[0m     X: DataFrame,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1378\u001b[0m     show_progress: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m   1379\u001b[0m ):\n\u001b[0;32m-> 1380\u001b[0m     beta_, ll_, hessian_ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_newton_raphson_for_efron_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1381\u001b[0m \u001b[43m        \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1382\u001b[0m \u001b[43m        \u001b[49m\u001b[43mT\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1383\u001b[0m \u001b[43m        \u001b[49m\u001b[43mE\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1384\u001b[0m \u001b[43m        \u001b[49m\u001b[43mweights\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1385\u001b[0m \u001b[43m        \u001b[49m\u001b[43mentries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1386\u001b[0m \u001b[43m        \u001b[49m\u001b[43minitial_point\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minitial_point\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1387\u001b[0m \u001b[43m        \u001b[49m\u001b[43mshow_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mshow_progress\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1388\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mfit_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1389\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1391\u001b[0m     \u001b[38;5;66;03m# compute the baseline hazard here.\u001b[39;00m\n\u001b[1;32m   1392\u001b[0m     predicted_partial_hazards_ \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m   1393\u001b[0m         pd\u001b[38;5;241m.\u001b[39mDataFrame(np\u001b[38;5;241m.\u001b[39mexp(dot(X, beta_)), columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mP\u001b[39m\u001b[38;5;124m\"\u001b[39m])\u001b[38;5;241m.\u001b[39massign(T\u001b[38;5;241m=\u001b[39mT\u001b[38;5;241m.\u001b[39mvalues, E\u001b[38;5;241m=\u001b[39mE\u001b[38;5;241m.\u001b[39mvalues, W\u001b[38;5;241m=\u001b[39mweights\u001b[38;5;241m.\u001b[39mvalues)\u001b[38;5;241m.\u001b[39mset_index(X\u001b[38;5;241m.\u001b[39mindex)\n\u001b[1;32m   1394\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/envs/BENG285P1/lib/python3.10/site-packages/lifelines/fitters/coxph_fitter.py:1539\u001b[0m, in \u001b[0;36mSemiParametricPHFitter._newton_raphson_for_efron_model\u001b[0;34m(self, X, T, E, weights, entries, initial_point, show_progress, step_size, precision, r_precision, max_steps)\u001b[0m\n\u001b[1;32m   1534\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exceptions\u001b[38;5;241m.\u001b[39mConvergenceError(\n\u001b[1;32m   1535\u001b[0m \u001b[38;5;250m        \u001b[39m\u001b[38;5;124;03m\"\"\"Hessian or gradient contains nan or inf value(s). Convergence halted. {0}\"\"\"\u001b[39;00m\u001b[38;5;241m.\u001b[39mformat(CONVERGENCE_DOCS),\n\u001b[1;32m   1536\u001b[0m         e,\n\u001b[1;32m   1537\u001b[0m     )\n\u001b[1;32m   1538\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e, LinAlgError):\n\u001b[0;32m-> 1539\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exceptions\u001b[38;5;241m.\u001b[39mConvergenceError(\n\u001b[1;32m   1540\u001b[0m \u001b[38;5;250m        \u001b[39m\u001b[38;5;124;03m\"\"\"Convergence halted due to matrix inversion problems. Suspicion is high collinearity. {0}\"\"\"\u001b[39;00m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m   1541\u001b[0m             CONVERGENCE_DOCS\n\u001b[1;32m   1542\u001b[0m         ),\n\u001b[1;32m   1543\u001b[0m         e,\n\u001b[1;32m   1544\u001b[0m     )\n\u001b[1;32m   1545\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1546\u001b[0m     \u001b[38;5;66;03m# something else?\u001b[39;00m\n\u001b[1;32m   1547\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n",
      "\u001b[0;31mConvergenceError\u001b[0m: Convergence halted due to matrix inversion problems. Suspicion is high collinearity. Please see the following tips in the lifelines documentation: https://lifelines.readthedocs.io/en/latest/Examples.html#problems-with-convergence-in-the-cox-proportional-hazard-modelMatrix is singular."
     ]
    }
   ],
   "source": [
    "sig_uni = univariate_cox(df, data_type=\"mutation\", q_thresh=0.05)\n",
    "sig_multi = multivariate_cox(df, data_type=\"mutation\", q_thresh=0.05)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "c3aa1dfb-c9cd-4b26-979e-8168cb01d513",
   "metadata": {},
   "outputs": [],
   "source": [
    "from lifelines import CoxPHFitter\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def univariate_cox(df, data_type, q_thresh=0.05, min_patients=50):\n",
    "    results = []\n",
    "    for feature in features:\n",
    "        try:\n",
    "            # Only keep feature if ≥ N patients have mutation\n",
    "            if df[feature].sum() < min_patients:\n",
    "                continue\n",
    "\n",
    "            # Check for separation: low variance in one group\n",
    "            events = df[\"OS\"].astype(bool)\n",
    "            var1 = df.loc[events, feature].var()\n",
    "            var2 = df.loc[~events, feature].var()\n",
    "            if (var1 < 1e-4 or var2 < 1e-4):  # very low variance\n",
    "                continue\n",
    "\n",
    "            # Assemble data\n",
    "            data = df[[feature, 'OS_time', 'OS']].dropna()\n",
    "            cph = CoxPHFitter()\n",
    "            cph.fit(data, duration_col='OS_time', event_col='OS', show_progress=False)\n",
    "            summ = cph.summary.loc[feature]\n",
    "\n",
    "            results.append({\n",
    "                'feature':     feature,\n",
    "                'coef':        summ['coef'],\n",
    "                'exp(coef)':   summ['exp(coef)'],\n",
    "                'zscore':      summ['z'],\n",
    "                'p':           summ['p'],\n",
    "                'lower 95%':   summ['exp(coef) lower 95%'],\n",
    "                'upper 95%':   summ['exp(coef) upper 95%']\n",
    "            })\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"⚠️ Skipped {feature}: {str(e)}\")\n",
    "            continue\n",
    "\n",
    "    # Multiple testing correction\n",
    "    res_df = pd.DataFrame(results).set_index('feature')\n",
    "    if not res_df.empty:\n",
    "        res_df[\"q\"] = multipletests(res_df['p'], method='fdr_bh')[1]\n",
    "        res_df = res_df.sort_values('q')\n",
    "        res_df.to_csv(f'{res_dir}{data_type}_univariate_cox_results.tsv', sep=\"\\t\", index=True)\n",
    "        sig = res_df[res_df[\"q\"] < q_thresh]\n",
    "        sig.to_csv(f'{res_dir}{data_type}_univariate_cox_significant_features.tsv', sep=\"\\t\")\n",
    "        return sig\n",
    "    else:\n",
    "        print(\"No valid features after filtering.\")\n",
    "        return pd.DataFrame()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "906c160e-f002-44f2-bfa7-6dcbb8f999a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_uni = univariate_cox(df, data_type=\"mutation\", q_thresh=0.05, min_patients=50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06ea8843-78ed-4665-b4fd-3a3d3d3af754",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
